src/core/thread/link_quality.cpp

```diff
@@ -0,0 +1,274 @@
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements link quality information processing and storage.
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include <openthread-types.h>
+#include <common/code_utils.hpp>
+#include <thread/link_quality.hpp>
+
+namespace Thread {
+
+// This array gives the decimal point digits representing 0/8, 1/8, ..., 7/8 (it does not include the '.').
+static const char *const kLinkQualityDecimalDigitsString[8] =
+{
+    // 0/8, 1/8,   2/8,   3/8,   4/8,   5/8,   6/8,   7/8
+    "000", "125", "250", "375", "500", "625", "750", "875"
+};
+
+LinkQualityInfo LinkQualityInfo::sNoiseFloorAverage;
+
+const char LinkQualityInfo::kUnknownRssString[] = "Unknown RSS";
+
+//-------------------------------------------------------------------------------
+
+LinkQualityInfo::LinkQualityInfo(void)
+{
+    Clear();
+}
+
+void LinkQualityInfo::Clear(void)
+{
+    mRssAverage = 0;
+    mCount = 0;
+    mLinkQuality = 0;
+}
+
+ThreadError LinkQualityInfo::AddRss(int8_t anRss)
+{
+    ThreadError error = kThreadError_None;
+    uint16_t    newValue;
+    uint16_t    oldAverage;
+
+    VerifyOrExit(anRss < 0, error = kThreadError_InvalidArgs);
+    VerifyOrExit(anRss != kUnknownRss, error = kThreadError_InvalidArgs);
+
+    // Multiply the the RSS value by -8 (to make it positive add precision to average).
+
+    newValue = -anRss;
+    newValue <<= 3;
+
+    if (mCount < 8)
+    {
+        mCount++;
+    }
+
+    oldAverage = mRssAverage;
+
+    if (mCount >= 8)
+    {
+        // New average = old average * 7/8 + new value * 1/8
+        mRssAverage = ((oldAverage << 3) - oldAverage + newValue) >> 3;
+    }
+    else if (mCount >= 4)
+    {
+        // New average = old average * 3/4 + new value * 1/4
+        mRssAverage = ((oldAverage << 2) - oldAverage + newValue) >> 2;
+    }
+    else if (mCount >= 2)
+    {
+        // New average = old average * 1/2 + new value * 1/2
+        mRssAverage = (oldAverage + newValue) >> 1;
+    }
+    else // count is 1
+    {
+        mRssAverage = newValue;
+    }
+
+exit:
+
+    if (error == kThreadError_None)
+    {
+        UpdateLinkQuality();
+    }
+
+    return error;
+}
+
+int8_t LinkQualityInfo::GetAverageRss(void) const
+{
+    int8_t average = kUnknownRss;
+
+    if (mCount != 0)
+    {
+        average = -(static_cast<int8_t>(mRssAverage >> 3));
+    }
+
+    return average;
+}
+
+uint16_t LinkQualityInfo::GetAverageRssAsEncodedWord(void) const
+{
+    return mRssAverage;
+}
+
+ThreadError LinkQualityInfo::GetAverageRssAsString(char *aCharBuffer, size_t aBufferLen) const
+{
+    ThreadError error = kThreadError_None;
+    int charsWritten = 0;
+
+    if (mCount == 0)
+    {
+        VerifyOrExit(aBufferLen >= sizeof(kUnknownRssString), error = kThreadError_NoBufs);
+
+        strcpy(aCharBuffer, kUnknownRssString);
```

Comment:
We check the buffer len to see if we have space before `strcpy` but I agree that `strncpy` will still be safer. Will change it